File: src/main/java/ink/rubi/codegenerator/controller/GlobalController.java
package ink.rubi.codegenerator.controller;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.generator.config.rules.DateType;
import ink.rubi.codegenerator.App;
import ink.rubi.codegenerator.controller.converter.ItemStringConverter;
import ink.rubi.codegenerator.po.DateTypeItem;
import ink.rubi.codegenerator.po.IdTypeItem;
import ink.rubi.codegenerator.po.holder.GlobalConfigHolder;
import javafx.fxml.FXML;
import javafx.scene.control.CheckBox;
import javafx.scene.control.ChoiceBox;
import javafx.scene.control.TextField;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.GridPane;
import javafx.stage.DirectoryChooser;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

import java.io.File;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.ResourceBundle;

/**
 * @author : Rubi
 * @version : 2019-05-17 22:34 下午
 */
@Slf4j
@Getter
public class GlobalController implements IController<GlobalConfigHolder> {

    @FXML
    private TextField outputDir, author, entityName, mapperName, xmlName, serviceName, serviceImplName, controllerName;
    @FXML
    private CheckBox open, fileOverride, enableCache, kotlin, swagger2, activeRecord, baseResultMap, baseColumnList;
    @FXML
    private ChoiceBox<DateTypeItem> dateType;
    @FXML
    private ChoiceBox<IdTypeItem> idType;
    @FXML
    private GridPane globalPage;

    private MainController mainController;

    private DirectoryChooser chooseGenerateDirectory;


    private static final List<DateTypeItem> dateTypes = new ArrayList<DateTypeItem>() {{
        add(new DateTypeItem(DateType.ONLY_DATE, "只使用 java.util.date 代替"));
        add(new DateTypeItem(DateType.SQL_PACK, "使用 java.sql 包下的"));
        add(new DateTypeItem(DateType.TIME_PACK, "使用 java.time 包下的,java8 新的时间类型"));
    }};

    private static final List<IdTypeItem> idTypes = new ArrayList<IdTypeItem>() {{
        add(new IdTypeItem(IdType.AUTO, "数据库ID自增"));
        add(new IdTypeItem(IdType.NONE, "该类型为未设置主键类型"));
        add(new IdTypeItem(IdType.INPUT, "用户输入ID,可以通过自己注册自动填充插件进行填充"));
        add(new IdTypeItem(IdType.ID_WORKER, "全局唯一ID (idWorker)"));
        add(new IdTypeItem(IdType.UUID, "全局唯一ID (UUID)"));
        add(new IdTypeItem(IdType.ID_WORKER_STR, "字符串全局唯一ID (idWorker 的字符串表示)"));
    }};

    private static final DateTypeItem defaultDateType = dateTypes.get(0);
    private static final IdTypeItem defaultIdType = idTypes.get(0);


    @Override
    public void init(MainController mainController) {
        this.mainController = mainController;
    }

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        chooseGenerateDirectory = new DirectoryChooser();
        chooseGenerateDirectory.setTitle("选择导出目录");
        dateType.getItems().addAll(dateTypes);
        dateType.setValue(defaultDateType);
        dateType.converterProperty().set(new ItemStringConverter<>());
        idType.getItems().addAll(idTypes);
        idType.setValue(defaultIdType);
        idType.converterProperty().set(new ItemStringConverter<>());
    }

    public void choose(MouseEvent mouseEvent) {
        File file = chooseGenerateDirectory.showDialog(App.getWindow());
        if (file != null) {
            String path = file.getPath();
            outputDir.setText(path);
        }
    }


    @Override
    public GlobalConfigHolder getConfigHolder() {
        return new GlobalConfigHolder().setAuthor(author.getText())
                .setOutputDir(outputDir.getText())
                .setFileOverride(fileOverride.isSelected())
                .setEnableCache(enableCache.isSelected())
                .setActiveRecord(activeRecord.isSelected())
                .setKotlin(kotlin.isSelected())
                .setSwagger2(swagger2.isSelected())
                .setBaseResultMap(baseResultMap.isSelected())
                .setBaseColumnList(baseColumnList.isSelected())
                .setEntityName(entityName.getText())
                .setMapperName(mapperName.getText())
                .setXmlName(xmlName.getText())
                .setServiceName(serviceName.getText())
                .setServiceImplName(serviceImplName.getText())
                .setControllerName(controllerName.getText())
                .setIdType(idType.getSelectionModel().selectedItemProperty().getValue().getIdType().name())
                .setDateType(dateType.getSelectionModel().selectedItemProperty().getValue().getDateType().name())
                .setOpen(open.isSelected());

    }

    @Override
    public void flushConfig(GlobalConfigHolder holder) {
        outputDir.setText(holder.getOutputDir());
        author.setText(holder.getAuthor());
        entityName.setText(holder.getEntityName());
        mapperName.setText(holder.getMapperName());
        xmlName.setText(holder.getXmlName());
        serviceName.setText(holder.getServiceName());
        serviceImplName.setText(holder.getServiceImplName());
        controllerName.setText(holder.getControllerName());
        open.setSelected(holder.isOpen());
        fileOverride.setSelected(holder.isFileOverride());
        enableCache.setSelected(holder.isEnableCache());
        kotlin.setSelected(holder.isKotlin());
        swagger2.setSelected(holder.isSwagger2());
        activeRecord.setSelected(holder.isActiveRecord());
        baseResultMap.setSelected(holder.isBaseResultMap());
        baseColumnList.setSelected(holder.isBaseColumnList());
        matchChoice(dateType, DateType.class, holder.getDateType(), defaultDateType);
        matchChoice(idType, IdType.class, holder.getIdType(), defaultIdType);

    }
}

File: src/main/java/ink/rubi/coffee/controller/GlobalController.kt
package ink.rubi.coffee.controller

import com.baomidou.mybatisplus.annotation.IdType
import com.baomidou.mybatisplus.generator.config.rules.DateType
import ink.rubi.coffee.controller.converter.ItemStringConverter
import ink.rubi.coffee.po.DateTypeItem
import ink.rubi.coffee.po.IdTypeItem
import ink.rubi.coffee.po.holder.GlobalConfigHolder
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.control.CheckBox
import javafx.scene.control.ChoiceBox
import javafx.scene.control.TextField
import javafx.scene.input.MouseEvent
import javafx.scene.layout.GridPane
import javafx.stage.DirectoryChooser
import java.net.URL
import java.util.*

/**
 * @author : Rubi
 * @version : 2019-05-17 22:34 下午
 */
class GlobalController : IController<GlobalConfigHolder>, Initializable {

    @FXML
    private lateinit var outputDir: TextField
    @FXML
    private lateinit var author: TextField
    @FXML
    private lateinit var entityName: TextField
    @FXML
    private lateinit var mapperName: TextField
    @FXML
    private lateinit var xmlName: TextField
    @FXML
    private lateinit var serviceName: TextField
    @FXML
    private lateinit var serviceImplName: TextField
    @FXML
    private lateinit var controllerName: TextField
    @FXML
    private lateinit var open: CheckBox
    @FXML
    private lateinit var fileOverride: CheckBox
    @FXML
    private lateinit var enableCache: CheckBox
    @FXML
    private lateinit var kotlin: CheckBox
    @FXML
    private lateinit var swagger2: CheckBox
    @FXML
    private lateinit var activeRecord: CheckBox
    @FXML
    private lateinit var baseResultMap: CheckBox
    @FXML
    private lateinit var baseColumnList: CheckBox
    @FXML
    private lateinit var dateType: ChoiceBox<DateTypeItem>
    @FXML
    private lateinit var idType: ChoiceBox<IdTypeItem>
    @FXML
    private lateinit var globalPage: GridPane

    private lateinit var mainController: MainController

    private lateinit var chooseGenerateDirectory: DirectoryChooser

    override fun inject(mainController: MainController) {
        this.mainController = mainController
    }

    override fun defaultShow() {
        outputDir.text = "D:\\generate-code"
        author.clear()
        entityName.clear()
        mapperName.clear()
        xmlName.clear()
        serviceName.clear()
        serviceImplName.clear()
        controllerName.clear()
        open.isSelected = true
        fileOverride.isSelected = false
        enableCache.isSelected = false
        kotlin.isSelected = false
        swagger2.isSelected = true
        activeRecord.isSelected = false
        baseResultMap.isSelected = true
        baseColumnList.isSelected = true
        dateType.value = defaultDateType
        idType.value = defaultIdType
    }

    override fun initialize(location: URL, resources: ResourceBundle?) {
        chooseGenerateDirectory = DirectoryChooser()
        chooseGenerateDirectory.title = "选择导出目录"
        dateType.items.addAll(dateTypes)
        dateType.converterProperty().set(ItemStringConverter())
        idType.items.addAll(idTypes)
        idType.converterProperty().set(ItemStringConverter())
        defaultShow()
    }

    fun choose(mouseEvent: MouseEvent) {
        val file = chooseGenerateDirectory.showDialog(mainController.window)
        if (file != null) {
            val path = file.path
            outputDir.text = path
        }
    }


    override fun getConfigHolder(): GlobalConfigHolder {
        val holder = GlobalConfigHolder()
        holder.author = author.text
        holder.outputDir = outputDir.text
        holder.isFileOverride = fileOverride.isSelected
        holder.isEnableCache = enableCache.isSelected
        holder.isActiveRecord = activeRecord.isSelected
        holder.isKotlin = kotlin.isSelected
        holder.isSwagger2 = swagger2.isSelected
        holder.isBaseResultMap = baseResultMap.isSelected
        holder.isBaseColumnList = baseColumnList.isSelected
        holder.entityName = entityName.text
        holder.mapperName = mapperName.text
        holder.xmlName = xmlName.text
        holder.serviceName = serviceName.text
        holder.serviceImplName = serviceImplName.text
        holder.controllerName = controllerName.text
        holder.idType = idType.selectionModel.selectedItemProperty().value.value?.name
        holder.dateType = dateType.selectionModel.selectedItemProperty().value.value!!.name
        holder.isOpen = open.isSelected
        return holder
    }

    override fun flushConfig(holder: GlobalConfigHolder) {
        outputDir.text = holder.outputDir
        author.text = holder.author
        entityName.text = holder.entityName
        mapperName.text = holder.mapperName
        xmlName.text = holder.xmlName
        serviceName.text = holder.serviceName
        serviceImplName.text = holder.serviceImplName
        controllerName.text = holder.controllerName
        open.isSelected = holder.isOpen
        fileOverride.isSelected = holder.isFileOverride
        enableCache.isSelected = holder.isEnableCache
        kotlin.isSelected = holder.isKotlin
        swagger2.isSelected = holder.isSwagger2
        activeRecord.isSelected = holder.isActiveRecord
        baseResultMap.isSelected = holder.isBaseResultMap
        baseColumnList.isSelected = holder.isBaseColumnList
        matchChoice(dateType, DateType::class.java, holder.dateType, defaultDateType)
        matchChoice(idType, IdType::class.java, holder.idType!!, defaultIdType)

    }

    companion object {


        private val dateTypes = object : ArrayList<DateTypeItem>() {
            init {
                add(DateTypeItem(DateType.ONLY_DATE, "只使用 java.util.date 代替"))
                add(DateTypeItem(DateType.SQL_PACK, "使用 java.sql 包下的"))
                add(DateTypeItem(DateType.TIME_PACK, "使用 java.time 包下的,java8 新的时间类型"))
            }
        }

        private val idTypes = object : ArrayList<IdTypeItem>() {
            init {
                add(IdTypeItem(IdType.AUTO, "数据库ID自增"))
                add(IdTypeItem(IdType.NONE, "该类型为未设置主键类型"))
                add(IdTypeItem(IdType.INPUT, "用户输入ID,可以通过自己注册自动填充插件进行填充"))
                add(IdTypeItem(IdType.ID_WORKER, "全局唯一ID (idWorker)"))
                add(IdTypeItem(IdType.UUID, "全局唯一ID (UUID)"))
                add(IdTypeItem(IdType.ID_WORKER_STR, "字符串全局唯一ID (idWorker 的字符串表示)"))
            }
        }

        private val defaultDateType = dateTypes[0]
        private val defaultIdType = idTypes[0]
    }
}
